nameserver/nameserver.go

```diff
@@ -25,34 +26,109 @@ const (
 
 	// Used by prog/weaver/main.go and proxy/create_container_interceptor.go
 	DefaultDomain = "weave.local."
+
+	// Used as a key for stored nameserver entries in the database
+	nameserverIdent = "nameserver"
+
+	// Used as an identifier for entries which do not belong to the Weave network
+	externContainerID = "weave:extern"
 )
 
 // Nameserver: gossip-based, in memory nameserver.
 // - Holds a sorted list of (hostname, peer, container id, ip) tuples for the whole cluster.
+// - DB handle for persisting local entries to survive container/peer restarts.
 // - This list is gossiped & merged around the cluser.
 // - Lookup-by-hostname are O(nlogn), and return a (copy of a) slice of the entries
 // - Update is O(n) for now
 type Nameserver struct {
 	sync.RWMutex
-	ourName     mesh.PeerName
-	domain      string
-	gossip      mesh.Gossip
-	entries     Entries
-	isKnownPeer func(mesh.PeerName) bool
-	quit        chan struct{}
+	ourName          mesh.PeerName
+	domain           string
+	gossip           mesh.Gossip
+	entries          Entries
+	isKnownPeer      func(mesh.PeerName) bool
+	quit             chan struct{}
+	db               db.DB
+	pendingBroadcast Entries
 }
 
-func New(ourName mesh.PeerName, domain string, isKnownPeer func(mesh.PeerName) bool) *Nameserver {
-	return &Nameserver{
+func New(ourName mesh.PeerName, domain string, db db.DB,
+	isKnownPeer func(mesh.PeerName) bool) *Nameserver {
+
+	n := &Nameserver{
 		ourName:     ourName,
 		domain:      dns.Fqdn(domain),
+		db:          db,
 		isKnownPeer: isKnownPeer,
 		quit:        make(chan struct{}),
 	}
+	n.restoreFromDB()
+
+	return n
+}
+
+// Restores local entries from the database.
+// NB: called only during the initialization, thus the lock is not taken.
+func (n *Nameserver) restoreFromDB() {
+	var entries Entries
+
+	n.debugf("restore: starting...")
+	ok, err := n.db.Load(nameserverIdent, &n.entries)
+
+	switch {
+	case err != nil:
+		n.errorf("restore: cannot retrieve entries due to: %s", err)
+		return
+	case !ok:
+		n.infof("restore: cannot find any entry in DB")
+		return
+	}
+
+	// lHostname gets lost, because the gob encoder which is used internally by the DB
+	// wrapper does not serialize private fields, so we need to set this field manually
```

Comment:
Not serialising the lower-case version is intentional; it can be recovered from the real hostname so no point using up bytes on it.  I.e. the important thing is the decision not to serialise it, not the gob encoder.